### ISSUE TITLE
Make Uri parser more restrictive

### ISSUE THREAD

**Issue by asoltysik** [pepegar/hammock#113]
Currently it allows for some invalid URIs, which lead to runtime errors when passed to Apache HTTP (thrown by java.net.URI) or AsyncHttpClient.

Examples of invalid URIs which pass `hammock.Uri.fromString` parser:
`::`
`^&%$#@!`
`http://^&%$#@!/foo`

A couple of ideas for how to solve this:

- Wrap and use `java.net.URI` constructor as a parser, in type-safe way (probably the easiest)
- Use a library like: https://github.com/lemonlabsuk/scala-uri
- Improve existing parser/write/port a parser from scratch (https://tools.ietf.org/html/rfc3986)

**Comment by miciek** [comment]
@asoltysik Thanks! I think second option would be the best. I haven't used at `scala-uri` but it looks like it's way better than `java.net.URI`.

**Comment by asoltysik** [comment]
Looking at `scala-uri` it uses `parboiled`, which is a java library with scala wrapper. IIRC that makes it not an option as hammock is cross-published to ScalaJS.

**Comment by pepegar** [comment]
Yeah, the idea behind creating our own was to have something that worked for both scalaJS & JVM (and potentially scala native)

**Comment by miciek** [comment]
OK, @asoltysik I think that the only way forward then is to improve existing parser :)
